File: rust/src/providers/generic.rs
//! Generic, descriptor-driven search provider.
//!
//! A single provider implementation that can speak to any engine described by a
//! catalog [`EngineDescriptor`](super::engines::EngineDescriptor). This keeps
//! the fetch/normalize/error plumbing in one place while each engine only
//! declares its URL, request kind, and parser. Mirrors the JavaScript
//! `src/providers/generic.js` (issue #3 parity requirement).

use async_trait::async_trait;
use std::collections::BTreeMap;

use super::base::{SearchOptions, SearchProvider, SearchResult};
use super::engines::{EngineDescriptor, EngineKind, HttpMethod};
use crate::error::SearchError;
use crate::transport::{ReqwestTransport, SearchTransport, TransportRequest};

const USER_AGENT: &str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 \
                          (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

/// Descriptor-driven provider.
pub struct GenericProvider {
    descriptor: EngineDescriptor,
    enabled: bool,
    weight: f64,
}

impl GenericProvider {
    /// Create a new provider for the given engine descriptor.
    pub fn new(descriptor: EngineDescriptor) -> Self {
        Self {
            descriptor,
            enabled: true,
            weight: 1.0,
        }
    }

    /// The engine descriptor backing this provider.
    pub fn descriptor(&self) -> &EngineDescriptor {
        &self.descriptor
    }

    fn build_headers(&self, options: &SearchOptions) -> BTreeMap<String, String> {
        let mut headers = BTreeMap::new();
        headers.insert("User-Agent".to_string(), USER_AGENT.to_string());
        headers.insert("Accept-Language".to_string(), "en-US,en;q=0.9".to_string());
        let accept = match self.descriptor.kind {
            EngineKind::Json => "application/json",
            EngineKind::Text | EngineKind::Html => {
                "text/html,application/xhtml+xml,application/xml;q=0.9"
            }
        };
        headers.insert("Accept".to_string(), accept.to_string());

        if let Some(extra) = self.descriptor.headers {
            for (name, value) in extra(options) {
                headers.insert(name, value);
            }
        }
        headers
    }
}

#[async_trait]
impl SearchProvider for GenericProvider {
    fn name(&self) -> &str {
        self.descriptor.id
    }

    fn is_available(&self) -> bool {
        self.enabled
    }

    fn weight(&self) -> f64 {
        self.weight
    }

    fn set_weight(&mut self, weight: f64) {
        self.weight = weight.clamp(0.0, 1.0);
    }

    fn set_enabled(&mut self, enabled: bool) {
        self.enabled = enabled;
    }

    async fn search(
        &self,
        query: &str,
        options: &SearchOptions,
    ) -> Result<Vec<SearchResult>, SearchError> {
        self.search_with_transport(query, options, &ReqwestTransport::default())
            .await
    }

    async fn search_with_transport(
        &self,
        query: &str,
        options: &SearchOptions,
        transport: &dyn SearchTransport,
    ) -> Result<Vec<SearchResult>, SearchError> {
        if query.trim().is_empty() {
            return Ok(Vec::new());
        }

        let d = &self.descriptor;
        let limit = options.limit.unwrap_or(10);
        let url = (d.build_url)(query, options);
        let mut headers = self.build_headers(options);

        let method = match d.method {
            HttpMethod::Get => "GET",
            HttpMethod::Post => "POST",
        };
        let mut body = None;
        if let (HttpMethod::Post, Some(build_body)) = (d.method, d.build_body) {
            headers.insert(
                "Content-Type".to_string(),
                "application/x-www-form-urlencoded".to_string(),
            );
            body = Some((build_body)(query, options).into_bytes());
        }
        let response = transport
            .execute(TransportRequest {
                method: method.to_string(),
                url,
                headers,
                body,
            })
            .await?;

        if !(200..300).contains(&response.status) {
            return Err(SearchError::ApiError {
                provider: d.id.to_string(),
                message: format!("HTTP {}", response.status),
            });
        }
        let body = String::from_utf8_lossy(&response.body);
        Ok((d.parse)(&body, limit, options))
    }
}

File: js/src/providers/generic.js
/**
 * Generic, descriptor-driven search provider.
 *
 * A single provider implementation that can speak to any engine described by a
 * catalog descriptor (see {@link module:providers/api-engines} and
 * {@link module:providers/html-engines}). This keeps the fetch/normalize/error
 * plumbing in one place while each engine only declares its URL, request kind,
 * and parser.
 *
 * @module providers/generic
 */

import { BaseSearchProvider } from './base.js';
import { request } from '../transport.js';

const USER_AGENT =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 ' +
  '(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36';

/**
 * @typedef {Object} EngineDescriptor
 * @property {string} id - Stable provider id
 * @property {string} label - Human-readable label
 * @property {'search'|'knowledge'|'papers'|'code'} category - Provider category
 * @property {'json'|'text'|'html'} kind - How the response body is decoded
 * @property {boolean} [corsReadable] - Whether the endpoint is browser-CORS readable
 * @property {boolean} [defaultForCategory] - Default provider for its category
 * @property {'GET'|'POST'} [method] - HTTP method (default GET)
 * @property {(query: string, options: Object) => string} buildUrl - Build request URL
 * @property {(query: string, options: Object) => string} [buildBody] - Build POST body
 * @property {(options: Object) => Object} [headers] - Extra request headers
 * @property {(payload: *, limit: number, options: Object) => import('./base.js').SearchResult[]} parse - Parse response
 */

/**
 * Descriptor-driven provider.
 *
 * @extends BaseSearchProvider
 */
export class GenericProvider extends BaseSearchProvider {
  /**
   * @param {EngineDescriptor} descriptor - Engine descriptor
   * @param {Object} [config] - Provider configuration
   * @param {Function} [config.fetchImpl] - Injectable fetch implementation (for tests)
   */
  constructor(descriptor, config = {}) {
    super(descriptor.id);
    this.descriptor = descriptor;
    this.category = descriptor.category;
    this.label = descriptor.label;
    this.corsReadable = Boolean(descriptor.corsReadable);
    this.transport = config.transport || config.fetchImpl;
    this.fetchImpl = this.transport || globalThis.fetch;
  }

  /**
   * Perform a search against the descriptor's engine.
   *
   * @param {string} query - Search query
   * @param {import('./base.js').SearchOptions} [options] - Search options
   * @returns {Promise<import('./base.js').SearchResult[]>}
   */
  async search(query, options = {}) {
    if (!query || typeof query !== 'string' || query.trim().length === 0) {
      return [];
    }

    const limit = options.limit || 10;
    const d = this.descriptor;
    const url = d.buildUrl(query, options);
    const method = d.method || 'GET';

    const headers = {
      'User-Agent': USER_AGENT,
      'Accept-Language': 'en-US,en;q=0.9',
      ...(d.kind === 'json'
        ? { Accept: 'application/json' }
        : { Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9' }),
      ...(typeof d.headers === 'function' ? d.headers(options) : {}),
    };

    const init = { method, headers };
    if (method === 'POST' && typeof d.buildBody === 'function') {
      init.body = d.buildBody(query, options);
      headers['Content-Type'] = 'application/x-www-form-urlencoded';
    }

    try {
      const response = await request(this.transport, url, init, options);
      if (!response.ok) {
        throw new Error(`${d.id} returned status ${response.status}`);
      }
      const payload =
        d.kind === 'json' ? await response.json() : await response.text();
      return d.parse(payload, limit, options);
    } catch (error) {
      if (options.throwOnError) {
        throw error;
      }
      console.error(`${d.id} search error: ${error.message}`);
      return [];
    }
  }
}

/**
 * Create a {@link GenericProvider} from a descriptor.
 *
 * @param {EngineDescriptor} descriptor - Engine descriptor
 * @param {Object} [config] - Provider configuration
 * @returns {GenericProvider}
 */
export function createGenericProvider(descriptor, config) {
  return new GenericProvider(descriptor, config);
}
